File: src/test/java/com/mpe85/grampa/intercept/ReferenceRuleTest.java
package com.mpe85.grampa.intercept;

import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNotEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.concurrent.Callable;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.api.function.ThrowingSupplier;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import com.mpe85.grampa.intercept.RuleMethodInterceptor.ReferenceRule;
import com.mpe85.grampa.rule.Rule;
import com.mpe85.grampa.rule.RuleContext;
import com.mpe85.grampa.rule.impl.EmptyRule;

@ExtendWith(MockitoExtension.class)
public class ReferenceRuleTest {
	
	static final class Foo implements Callable<Rule<Integer>> {
		public Rule<Integer> rule1() {
			return new EmptyRule<>();
		}
		
		public Rule<Integer> rule2() {
			return new EmptyRule<>();
		}
		
		@Override
		public Rule<Integer> call() throws Exception {
			return new EmptyRule<>();
		}
	}
	
	@Test
	public void equalsHashCodeToString() {
		final ReferenceRule<Integer> rule1 = createReferenceRule("rule1");
		final ReferenceRule<Integer> rule2 = createReferenceRule("rule1");
		final ReferenceRule<Integer> rule3 = createReferenceRule("rule2");
		
		assertTrue(rule1.equals(rule2));
		assertFalse(rule1.equals(rule3));
		assertFalse(rule1.equals(new Object()));
		
		assertEquals(rule1.hashCode(), rule2.hashCode());
		assertNotEquals(rule1.hashCode(), rule3.hashCode());
		
		assertEquals(String.format("ReferenceRule{#children=0, hashCode=%d}", rule1.hashCode()), rule1.toString());
		assertEquals(String.format("ReferenceRule{#children=0, hashCode=%d}", rule2.hashCode()), rule2.toString());
		assertEquals(String.format("ReferenceRule{#children=0, hashCode=%d}", rule3.hashCode()), rule3.toString());
	}
	
	@Test
	public void match(@Mock final RuleContext<Integer> ctx) {
		final ReferenceRule<Integer> rule = createReferenceRule("rule1");
		
		assertFalse(rule.match(ctx));
	}
	
	private ReferenceRule<Integer> createReferenceRule(final String ruleMethod) {
		final RuleMethodInterceptor<Integer> interceptor = new RuleMethodInterceptor<>();
		
		final ThrowingSupplier<Rule<Integer>> supplier =
				() -> interceptor.intercept(Foo.class.getDeclaredMethod(ruleMethod), new Foo());
		
		assertDoesNotThrow(supplier);
		final Rule<Integer> rule = assertDoesNotThrow(supplier);
		
		assertTrue(rule instanceof ReferenceRule);
		
		return (ReferenceRule<Integer>) rule;
	}
	
}

File: src/test/kotlin/com/mpe85/grampa/intercept/ReferenceRuleTest.kt
package com.mpe85.grampa.intercept

import com.mpe85.grampa.rule.ReferenceRule
import com.mpe85.grampa.rule.Rule
import com.mpe85.grampa.rule.RuleContext
import com.mpe85.grampa.rule.impl.EmptyRule
import java.util.concurrent.Callable
import org.junit.jupiter.api.Assertions.assertDoesNotThrow
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.api.TestInstance.Lifecycle.PER_CLASS
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.api.function.ThrowingSupplier
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension

@TestInstance(PER_CLASS)
@ExtendWith(MockitoExtension::class)
class ReferenceRuleTest {

  internal class RuleMethods : Callable<Rule<Int>> {
    fun rule1(): Rule<Int> {
      return EmptyRule()
    }

    fun rule2(): Rule<Int> {
      return EmptyRule()
    }

    @Throws(Exception::class)
    override fun call(): Rule<Int> {
      return EmptyRule()
    }
  }

  @Test
  fun equalsHashCodeToString() {
    val rule1 = createReferenceRule("rule1")
    val rule2 = createReferenceRule("rule1")
    val rule3 = createReferenceRule("rule2")
    assertEquals(rule1, rule2)
    assertNotEquals(rule1, rule3)
    assertNotEquals(rule1, Any())
    assertEquals(rule1.hashCode(), rule2.hashCode())
    assertNotEquals(rule1.hashCode(), rule3.hashCode())
    assertEquals(
      String.format("ReferenceRule{#children=0, hashCode=%d}", rule1.hashCode()),
      rule1.toString()
    )
    assertEquals(
      String.format("ReferenceRule{#children=0, hashCode=%d}", rule2.hashCode()),
      rule2.toString()
    )
    assertEquals(
      String.format("ReferenceRule{#children=0, hashCode=%d}", rule3.hashCode()),
      rule3.toString()
    )
  }

  @Test
  fun match(@Mock ctx: RuleContext<Int>?) {
    val rule = createReferenceRule("rule1")
    assertFalse(rule.match(ctx))
  }

  private fun createReferenceRule(ruleMethod: String): ReferenceRule<Int> {
    val interceptor = RuleMethodInterceptor<Int>()
    val supplier = ThrowingSupplier {
      interceptor.intercept(
        RuleMethods::class.java.getDeclaredMethod(ruleMethod), RuleMethods()
      )
    }
    assertDoesNotThrow(supplier)
    val rule = assertDoesNotThrow(supplier)
    assertTrue(rule is ReferenceRule<*>)
    return rule as ReferenceRule<Int>
  }

}
